dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h

```diff
@@ -39,172 +39,263 @@ DALI_HOST_DEV DALI_FORCEINLINE vec<N, float> norm(vec<N, float> x) {
+    float y = dot(coeffs, rgb);
+    return needs_clamp<Input, Output>::value ? Convert<Output>(y) : ConvertSat<Output>(y);
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr Output rgb_to_cb(vec<3, Input> rgb) {
+    constexpr vec3 coeffs = vec3(-0.16873589f, -0.33126411f, 0.5f)
+                            * detail::scale_factor<Input, Output>();
+    constexpr float bias = 0.5f * detail::bias_scale<Output>();
+    float y = dot(coeffs, rgb) + bias;
+    return needs_clamp<Input, Output>::value ? Convert<Output>(y) : ConvertSat<Output>(y);
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr Output rgb_to_cr(vec<3, Input> rgb) {
+    constexpr vec3 coeffs = vec3(0.5f, -0.41868759f, -0.08131241f)
+                            * detail::scale_factor<Input, Output>();
+    constexpr float bias = 0.5f * detail::bias_scale<Output>();
+    float y = dot(coeffs, rgb) + bias;
+    return needs_clamp<Input, Output>::value ? Convert<Output>(y) : ConvertSat<Output>(y);
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr vec<3, Output> rgb_to_ycbcr(vec<3, Input> rgb) {
+    return {
+      rgb_to_y<Output, Input>(rgb),
+      rgb_to_cb<Output, Input>(rgb),
+      rgb_to_cr<Output, Input>(rgb)
+    };
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr vec<3, Output> ycbcr_to_rgb(vec<3, Input> ycbcr) {
+    constexpr float cbias = 0.5f * detail::bias_scale<Input>();
+    constexpr float s = detail::scale_factor<Input, Output>();
+    float tmp_b = ycbcr[1] - cbias;
+    float tmp_r = ycbcr[2] - cbias;
+    float ys = ycbcr[0] * s;
+    auto r = ConvertSat<Output>(ys + (1.402f * s) * tmp_r);
+    auto g = ConvertSat<Output>(ys - (0.344136285f * s) * tmp_b - (0.714136285f * s) * tmp_r);
+    auto b = ConvertSat<Output>(ys + (1.772f * s) * tmp_b);
+    return {r, g, b};
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr Output gray_to_y(Input gray) {
+    return ConvertSatNorm<Output>(gray);
+  }
+
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE
+  constexpr Output y_to_gray(Input y) {
+    return ConvertSatNorm<Output>(y);
+  }
+};  // struct jpeg
 
-template <typename Output, typename Input>
-DALI_HOST_DEV DALI_FORCEINLINE Output rgb_to_y(vec<3, Input> rgb_in) {
-  constexpr vec3 coeffs(0.299f, 0.587f, 0.114f);
-  auto rgb = detail::norm(rgb_in);  // TODO(janton): optimize number of multiplications
-  return ConvertSatNorm<Output>(dot(coeffs, rgb));
-}
 
 template <>
-DALI_HOST_DEV DALI_FORCEINLINE uint8_t rgb_to_y(vec<3, uint8_t> rgb) {
-  constexpr vec3 coeffs(0.299f, 0.587f, 0.114f);
+DALI_HOST_DEV DALI_FORCEINLINE
+uint8_t jpeg::rgb_to_y<uint8_t, uint8_t>(vec<3, uint8_t> rgb) {
```

Comment:
Is this specialization necessary?

Reply:
Perhaps not.

Reply:
And are we good with keeping the unnecessary ones? A downside is that it may be consufing for any code readers as to why given case needs special treatment.

Reply:
Removed.